File: vdslib/src/vespa/vdslib/state/nodetype.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "nodetype.h"
#include <vespa/vespalib/util/exceptions.h>
#include <vespa/vespalib/stllike/asciistream.h>
#include <cassert>

namespace storage::lib {

// WARNING: Because static initialization happen in random order, the State
// class use these enum values directly during static initialization since
// these objects may yet not exist. Update in State if you change this.
const NodeType NodeType::STORAGE("storage", NodeType::Type::STORAGE);
const NodeType NodeType::DISTRIBUTOR("distributor", NodeType::Type::DISTRIBUTOR);

const NodeType&
NodeType::get(vespalib::stringref serialized)
{
    if (serialized == STORAGE._name) {
        return STORAGE;
    }
    if (serialized == DISTRIBUTOR._name) {
        return DISTRIBUTOR;
    }
    throw vespalib::IllegalArgumentException(
            "Unknown node type " + serialized + " given.", VESPA_STRLOC);
}

const NodeType&
NodeType::get(Type type) noexcept
{
    switch (type) {
        case Type::STORAGE:
            return STORAGE;
        case Type::DISTRIBUTOR:
            return DISTRIBUTOR;
        case Type::UNKNOWN:
            assert(type != Type::UNKNOWN);
    }
    abort();
}

NodeType::NodeType(vespalib::stringref name, Type type) noexcept
    : _type(type), _name(name)
{
}

std::ostream & operator << (std::ostream & os, const NodeType & n) {
    return os << n.toString();
}

vespalib::asciistream & operator << (vespalib::asciistream & os, const NodeType & n) {
    return os << n.toString();
}

}

File: vdslib/src/main/java/com/yahoo/vdslib/state/NodeType.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vdslib.state;

import java.util.List;

public enum NodeType {
    STORAGE("storage"),
    DISTRIBUTOR("distributor");

    private final String serializeAs;

    NodeType(String serializeAs) {
        this.serializeAs = serializeAs;
    }

    public String toString() {
        return serializeAs;
    }

    public static NodeType get(String serialized) {
        for (NodeType type : values()) {
            if (type.serializeAs.equals(serialized)) return type;
        }
        throw new IllegalArgumentException("Unknown node type '" + serialized + "'. Legal values are 'storage' and 'distributor'.");
    }

    public static List<NodeType> getTypes() { return List.of(STORAGE, DISTRIBUTOR); }

}
